### DISCUSSION TITLE
Ignore a crate from the output

### DISCUSSION THREAD

**Discussion by AntoniosBarotsis** [xd009642/tarpaulin#1181] [Q&A] [answered]
I have a workspace with multiple crates and I want one of them to not be included in the coverage calculations.

I tried doing something `--exclude-files cratename/*.rs` but that did not work, is there any way around this?

**Comment by xd009642** [comment, marked as answer]
You can use `--exclude` to exclude a package, similar to cargo test in a workspace. `--exclude-files cratename/*` is probably a solution as well using --exclude-files but the crates tests will still run if it's in a workspace so code it calls in another workspace crate will be in the stats

**Comment by AntoniosBarotsis** [comment]
> The crate that I want to exclude is named `cli`.

I tried running

```
cargo tarpaulin --out html --workspace --exclude cli
```

Which worked but did **not** ignore `cli` in the coverage. 

I then tried removing the workspace

```
cargo tarpaulin --out html --exclude cli
```

which doesn't work since `error: --exclude can only be used together with --workspace`.

```
cargo tarpaulin --out html --workspace --exclude-files cli/*
```

did not work either.

Any ideas?

**Reply by xd009642** [reply]
Do you have a link to the project i can look at? Also which OS is it? Windows can be funny with paths so there are some things to fix there

**Reply by AntoniosBarotsis** [reply]
I am indeed on windows, the project is [here](https://github.com/AntoniosBarotsis/qr-rs/tree/dev). 

What issues are there with windows paths? Maybe I could help.

**Reply by xd009642** [reply]
Generally it's been different things with UNC paths, terminal envs causing paths to look something like `//?/C:/`, this form of the paths doesn't really work properly in a lot of places (inside the compiler, some parts of standard library i.e. canonicalise), and likely is broken in other libraries - so potentially file globs are broken. 

Oh looking at my glob code it uses regex not even an external library to do it so it's definitely me. Probably separators don't match when looking at paths vs globs etc. Should probably change it to use a glob library instead and add more tests around the area

**Comment by xd009642** [comment]
So moving to the glob crate there's one caveat (which is mildly irritating me but makes sense), the path separators need to match with the glob expression and path checked against. I have a branch with this in `fix/windows-glob` I'm just testing it on a windows machine rn.

EDIT: Scratch that it seems to work with both separators... Wonder why my test failed then :thinking:

**Reply by AntoniosBarotsis** [reply]
I'm not sure I follow 😅

**Reply by xd009642** [reply]
Don't worry about it, I just thought when testing `--exclude-files src/*` would exclude only paths tested like `src/foo.rs` and not `src\\foo.rs`. Because of a test I wrote and ran on linux but then it worked for both on windows. 

I do see the crate has some cfg(windows) code so that may be why - or I did something dumb :thinking:

**Comment by AntoniosBarotsis** [comment]
This is now fixed in the `dev` branch, probably in this commit 1f40d9a84b7626f6205e63da61be2f7022738d03. Thanks!
